src/strkey.rs

```diff
@@ -82,17 +98,172 @@ impl StrkeyPrivateKeyEd25519 {
+        match <[u8; 40]>::try_from(payload) {
+            Ok(muxed) => {
+                let (ed25519, id) = muxed.split_at(32);
+                Ok(Self {
+                    ed25519: ed25519.try_into().map_err(|_| DecodeError::Invalid)?,
+                    id: u64::from_be_bytes(id.try_into().map_err(|_| DecodeError::Invalid)?),
+                })
+            }
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::MUXED_ACCOUNT_ED25519 => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyPreAuthTx(pub [u8; 32]);
+
+impl StrkeyPreAuthTx {
+    pub fn to_string(&self) -> String {
+        encode(version::PRE_AUTH_TX, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(pre_auth_tx) => Ok(Self(pre_auth_tx)),
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::PRE_AUTH_TX => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeyHashX(pub [u8; 32]);
+
+impl StrkeyHashX {
+    pub fn to_string(&self) -> String {
+        encode(version::HASH_X, &self.0)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match payload.try_into() {
+            Ok(hash_x) => Ok(Self(hash_x)),
+            Err(_) => Err(DecodeError::Invalid),
+        }
+    }
+
+    pub fn from_string(s: &str) -> Result<Self, DecodeError> {
+        let (ver, payload) = decode(s)?;
+        match ver {
+            version::HASH_X => Self::from_payload(&payload),
+            _ => Err(DecodeError::Invalid),
+        }
+    }
+}
+
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct StrkeySignedPayloadEd25519 {
+    pub ed25519: [u8; 32],
+    pub payload: Vec<u8>,
+}
+
+impl StrkeySignedPayloadEd25519 {
+    pub fn to_string(&self) -> String {
+        let inner_payload_len = self.payload.len();
+        let payload_len = 32 + 4 + inner_payload_len + (4 - inner_payload_len % 4) % 4;
+
+        let mut payload = vec![0; payload_len];
+        payload[..32].copy_from_slice(&self.ed25519);
+        payload[32..32 + 4].copy_from_slice(&(inner_payload_len as u32).to_be_bytes());
+        payload[32 + 4..32 + 4 + inner_payload_len].copy_from_slice(&self.payload);
+
+        encode(version::SIGNED_PAYLOAD_ED25519, &payload)
+    }
+
+    fn from_payload(payload: &[u8]) -> Result<Self, DecodeError> {
+        match Vec::try_from(payload) {
```

Comment:
Conversion from `&[u8]` to `Vec` should never fail. What's the reason for it being a try_from that can fail with an error case?

Reply:
fixed a846eeb9a87ff463ba9daf9b6748c07d62bcf125